prow/git/git.go

```diff
@@ -310,22 +312,33 @@ func (r *Repo) CheckoutPullRequest(number int) error {
 // Config runs git config.
 func (r *Repo) Config(key, value string) error {
 	r.logger.Infof("Running git config %s %s", key, value)
-	if b, err := r.gitCommand("config", key, value).CombinedOutput(); err != nil {
+	if b, err := r.gitCommand("config", key, value); err != nil {
 		return fmt.Errorf("git config %s %s failed: %v. output: %s", key, value, err, string(b))
 	}
 	return nil
 }
 
+// Remote builds a remote url from user, pasword, and a slice of path items
+func Remote(base *url.URL, user string, pass string, pathItems ...string) *url.URL {
+	copy := *base
+	if user != "" && pass != "" {
```

Comment:
This cannot remain in our client

Reply:
I do not understand how and why this is offensive...

Reply:
We should not ever allow production code to have specific callouts for test fakes. This is what you're looking at and why the integration tests we have did not catch the issue that was created.

Reply:
Could you please elaborate?

Reply:
The reason none of our integration tests caught the error is that they pass no credentials, which is impossible in production. The tests and our production client use different code paths. That is not acceptable. Look at my refactor to see how we can construct the full URL before creating the client and ever leak it. It is similarly invalid to:

 - have a client but never call SetCredentials()
 - have a client but not call getCredentials() when constructing the remote
 - have a client but not call Remote() when using it for reaching out to the remote

We should not have a client whose structure makes _any_ of those programmer errors possible, because they are invalid as options and only increase our surface area for bugs.

Reply:
the problem was all tests were using no credentials, so the code path with credentials was not ever executed in any test. Question of API design is irrelevant for this matter.

Reply:
Two statements:

1. different code-paths for no credential case and for testing is unacceptable
2. this is similar to N other things which are also unacceptable about the client as it is today.

I'm not saying we need to tackle them all, you just asked for me to elaborate.